### ISSUE TITLE
Next approach to persistence API design

### ISSUE THREAD

**Issue by Horusiath** [Horusiath/Akkling#7]
Right one of the benefits of Akka F# API is the ability to represent actor as a function. This gives a great power for the functional programming as we have whole battery of mechanisms to work on functions. But... since original persistence API was based on 2 (or even 3) callbacks - ReceiveRecovery, ReceiveCommand and Persist callbacks - this was not a suitable model for FP. Right now Akklin.Persistence solves this without deep changing of the logic structure.

This was an old approach:

```fsharp
spawnPersist system "pid"
<| fun mailbox ->
    {
        apply: fun state e -> (* receive recover callback - compute new state and recover *)
        exec: fun msg -> 
            (* receive command callback *)
            match msg with
            | ToBePersisted(x) -> mailbox.Persist (fun e -> (* persist callback *)) [x]
            ...
    }
```

This was bad and irritating. Also way o solve immutable state updates was ugly. New approach is to combine all of those callbacks into one, just like in the original F# API:

```fsharp
spawnPersist system "pid"
<| fun mailbox ->
    let rec recovering state =
        actor {
            let! (msg: obj) = mailbox.Receive ()
            match msg with
            | :? SnapshotOffer offer -> return! recovering (offer.Snapshot)
            | :? Event as e -> return! recovering (update state e)
            | :? RecoveryCompleted -> return! receive state
            | _ -> return! recovering state
        }
    and receive state = 
        actor {
            let! msg = mailbox.Receive ()
            match msg with
            | CommandMsg1 -> return! receive state
            | CommandMsg2(x) -> 
                persist! (recovering state) [x]  // need to think on this
        }
```

Mailbox could have some flags/unions to identify if it's a recovering, persistence ack or standard phase. Second option (better?): resolve the phase this using Active Patterns inside matching statement itself:

```fsharp
match msg with
| Persisted e -> (* persistent callback *)
| Recovered e -> (* recovery callback *)
| e -> (* standard receive pattern *)
```

This could be used fluently and can be an optional mechanism.

One thing to remember when implementing - actor context and sender can change, what makes them volatile when used inside persist callbacks. This issue should be solved.

**Comment by Horusiath** [comment]
Most of the persistent API is ready.(examples are [here](https://github.com/Horusiath/Akkling/blob/actor-expression-2.0/examples/persistence.fsx)). 

Missing part is active pattern support.

**Comment by Horusiath** [comment]
Closed with release 0.2
